### ISSUE TITLE
Improve NixOS error messages for systemd.services config

### ISSUE THREAD

**Issue by crabdancing** [NixOS/nixpkgs#275296]
### Describe the bug
Trying to make a systemd oneshot remain after exit, results in strange build errors:

```
       error: A definition for option `systemd.services.helloWorld.serviceConfig' is not of type `attribute set of (systemd option)'. Definition values:
       - In `/nix/store/7fkwd086mmlkpmsk1xyn23h4rlak8qrz-source/flake.nix':
           {
             RemainAfterExit = "yes";
             Restart = "on-error";
             Type = "oneshot";
           }
```

### Steps To Reproduce
```nix
  systemd.services.helloWorld = {
    script =
    ''
      echo "Hello World"
    '';
    enable = true;
    description = "example";
    wantedBy = [ "multi-user.target" ];
    serviceConfig = {
      Type = "oneshot";
      Restart = "on-error";
      RemainAfterExit = "yes";
    };
  };
```

### Expected behavior

Should be able to make a oneshot that remains after exit.

### Notify maintainers

@nbraud

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.2, NixOS, 23.11 (Tapir), 23.11.20231128.7c4c205`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `""`
 - nixpkgs: `/etc/nixPath`
```

**Comment by eclairevoyant** [comment]
The `Restart` value is invalid, see `man systemd.service` for valid values.

**Comment by alxpettit** [comment]
Ah, right you are. It's `on-failure`, not `on-error`.

**Comment by alxpettit** [comment]
I'm just curious... is there any way to get the Nix code spec of the schema to print what specifically the is wrong with it? Are there any features planned for this -- that is, better error messages when types do not match?

**Comment by lytedev** [comment]
I agree with above -- that would be very cool. For myself, I was able to troubleshoot by examining the relevant systemd man pages. In my case, I am configuring `systemd-networkd`, so `man systemd.network` had the fields and values well-documented. In this case, you could look at  `man systemd.service`.

**Comment by crabdancing** [comment]
It's kind of unfortunate because, although having a formalized schema is the 'correct' approach, and can in principle have huge advantages, the end result when doing that with Nix is often something harder to troubleshoot than if it just deployed the broken config, a lot of the time. :P

**Comment by eclairevoyant** [comment]
You're welcome to reopen the ticket if you want to track improving the error messages. I don't see any open issues for this specifically, nor for submodules overall.

**Comment by crabdancing** [comment]
@eclarevoyant Sure! :)

**Comment by SuperSandro2000** [comment]
I think we need to replace this line https://github.com/NixOS/nixpkgs/blob/81788f7597ea358312f10f38300753746b2a90e4/nixos/lib/systemd-unit-options.nix#L391 with a proper type aka a submodule.

**Comment by Iizuki** [comment]
Just providing another example how this error may be encountered:

As a nix newbie I ran into it in systemd-networkd configs. I had converted a network file that most certainly worked on a non-nixos box. After some double checking & trial and error I discovered my error was providing booleans as strings. E.g. `UseDomains = "true";` instead of `UseDomains = true;`.
